dnscache/manager.go

```diff
@@ -0,0 +1,118 @@
+package dnscache
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+
+	"github.com/TykTechnologies/tyk/log"
+)
+
+var (
+	logger = log.Get().WithField("prefix", "dnscache")
+)
+
+type DialContextFunc func(ctx context.Context, network, address string) (net.Conn, error)
+
+// IDnsCacheManager is an interface for abstracting interaction with dns cache. Implemented by DnsCacheManager
+type IDnsCacheManager interface {
+	InitDNSCaching(ttl, checkInterval time.Duration)
+	WrapDialer(dialer *net.Dialer) DialContextFunc
+	SetCacheStorage(cache IDnsCacheStorage)
+	CacheStorage() IDnsCacheStorage
+	DisposeCache()
+}
+
+// IDnsCacheStorage is an interface for working with cached storage of dns record.
+// Wrapped by IDnsCacheManager/DnsCacheManager. Implemented by DnsCacheStorage
+type IDnsCacheStorage interface {
+	FetchItem(key string) ([]string, error)
+	Get(key string) (DnsCacheItem, bool)
+	Set(key string, addrs []string)
+	Delete(key string)
+	Clear()
+}
+
+// DnsCacheManager is responsible for in-memory dns query records cache.
+// It allows to init dns caching and to hook into net/http dns resolution chain in order to cache query response ip records.
+type DnsCacheManager struct {
+	cacheStorage IDnsCacheStorage
+}
+
+// NewDnsCacheManager returns new empty/non-initialized DnsCacheManager
+func NewDnsCacheManager() *DnsCacheManager {
+	return &DnsCacheManager{nil}
+}
+
+func (m *DnsCacheManager) SetCacheStorage(cache IDnsCacheStorage) {
+	m.cacheStorage = cache
+}
+
+func (m *DnsCacheManager) CacheStorage() IDnsCacheStorage {
+	return m.cacheStorage
+}
+
+// WrapDialer returns wrapped version of net.Dialer#DialContext func with hooked up caching of dns queries.
+//
+// Actual dns server call occures in net.Resolver#LookupIPAddr method,
+// linked to net.Dialer instance by net.Dialer#Resolver field
+func (m *DnsCacheManager) WrapDialer(dialer *net.Dialer) DialContextFunc {
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		return m.doCachedDial(dialer, ctx, network, address)
+	}
+}
+
+func (m *DnsCacheManager) doCachedDial(d *net.Dialer, ctx context.Context, network, address string) (net.Conn, error) {
+	safeDial := func(addr string, itemKey string) (net.Conn, error) {
+		conn, err := d.DialContext(ctx, network, addr)
+		if err != nil && itemKey != "" {
+			m.cacheStorage.Delete(itemKey)
+		}
+		return conn, err
+	}
+
+	if m.cacheStorage == nil {
+		return safeDial(address, "")
+	}
+
+	host, port, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+
+	if ip := net.ParseIP(host); ip != nil {
+		return safeDial(address, "")
+	}
+
+	ips, err := m.cacheStorage.FetchItem(host)
```

Comment:
Why? 
We'll break the balancer logic in that case. 

In our case we'll always get a cached list and the connection will be established with the 1st list item/ip until cache for that record will expire and next dns server query return **probably** list with newly ordered elements.

Reply:
`in our case we'll always get a cached list and the connection will be established with the 1st list item/ip until cache for that record will expire` - that is my concern. With this implementation - no matter how DNS reply shuffles IPs we will be always bombing the last know/cached IP until it expires in cache. Why `**probably** list with reordered elements`. I think it is not probably, it is more likely as several A-records in DNS reply is the implementation of round robin on DNS level

Reply:
So, the suggestion is: 
a) to use cached ip only if there' exist 1 item for CNAME record and for `>1' case - fallback to native dial using `net.Dialer#DialContext(...address)`, which will [dialSerial](https://golang.org/src/net/dial.go#L406)
b)add a tyk's round robin on cached ips if we want to archieve a balancing on cached records(and therefor not to hit 1st only from list)
c) none of the above?

Reply:
not sure I follow here, what do you mean by `CNAME`? when we resolve hostname to addr we receive back IP, we don't know was it CNAME or A

Reply:
But this `PR` is created to fix caching for *some* servers, which doesn't support keep-alive or *do support it*(@buger FYI) and close/decide to close it when *tyk is initiating such connection* or *after it sends* response to tyk. 

Here's the original issue for details: 
`In the case when upstream does not support keep-alive connections, DNS query will be made on each HTTP request, which makes the huge load on DNS server.`

Reply:
>not sure I follow here, what do you mean by `CNAME`? when we resolve hostname to addr we receive back IP, we don't know was it CNAME or A
Sorry, by this 
>to use cached ip only if there' exist 1 item for CNAME record

I meant "to use cached ip only if there exist 1 ip after recursive dns resolution chain finishes".

Reply:
Using all IPs from the record is the safest approach. I vote to just fetch random IP record instead of the first one, which is kinda simplified round robin.